Reimplement replaceAllOccurrences and recipe printing

TheSilkMiner opened the pull request:
Reimplements the old `replaceAllOccurrences` in a new form, allowing for replacement of ingredients within single recipes, single recipe managers, or across the whole set of recipes, allowing for batching the replacements to avoid looping over the whole registry for every single recipe.
Also provides the recipe printing feature.

This is still early in development, and it's still in need of polishing:

- [x] Define a set of `ReplacementRule`s for vanilla types and dictate which rules are which
- [x] Refactor commands to avoid having extraneous functions in `CraftTweakerAPI`
- [x] Test
- [x] ...
- [x] Documentation

jaredlll08 committed:
Add RecipeWriting

jaredlll08 committed:
add /ct recipes hand

TheSilkMiner committed:
Begin moving from IRecipeWriter to IRecipeHandler

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Move recipe handlers to own registry

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Rewrite recipe handlers to account for current changes

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add Replacer

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add some more utility methods to replacement rule

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add cooking and stonecutter replacement handlers

These are still early on and feature repeated logic: IRecipeHandler may
provide a similar method as a static for ease of usage, if this is the
direction that we want to follow. This is merely to show off a potential
implementation.

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/CraftTweakerRegistry.java:
```diff
@@ -57,6 +61,8 @@ public static void findClasses() {
         ZEN_CLASS_REGISTRY.getImplementationsOf(TagManager.class)
                 .forEach(CrTTagRegistryData.INSTANCE::addTagImplementationClass);
+        
+        getAllClassesWith(IRecipeHandler.For.class).distinct().forEach(RECIPE_HANDLER_REGISTRY::addClass);
```
I don't quite remember, does this this find them if they use the container annotation?

Like, can you try both
```java
@Container({@For(...), @For(...)})
public class HandlerA{}

@For(...) @For(...)
public class HandlerB{}

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/IRecipeHandler.java:
```diff
@@ -0,0 +1,161 @@
+     * class.
+     */
+    default Optional<T> replaceIngredients(final IRecipeManager manager, final T recipe, final List<ReplacementRule> rules) {
+        throw new UnsupportedOperationException("Replacement is not supported for this recipe type");
```
Instead of throwing we could make it return `Optional.empty()` by defaullt?
Otherwise, how do you separate "not supported" from "some error occured while replacing"?

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/IRecipeHandler.java:
```diff
@@ -0,0 +1,161 @@
+     * @param recipe The recipe that is currently being dumped.
+     * @return A String representing a {@code addRecipe} (or similar) call.
+     */
+    String dumpToCommandString(final IRecipeManager manager, final T recipe);
```
More out of interest:

Why do we need the manager here?
Wouldn't the recipe be enough in most cases?

Or are there some cases where one implementation of a recipe class would be in more than one registry?

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/RecipeHandlerRegistry.java:
```diff
@@ -0,0 +1,89 @@
+    private final Map<Class<? extends IRecipe<?>>, IRecipeHandler<?>> recipeHandlers = new HashMap<>();
+    
+    public void addClass(final Class<?> clazz) {
+        if (!IRecipeHandler.class.isAssignableFrom(clazz)) {
+            throw new IllegalArgumentException("Class " + clazz.getName() + " does not implement IRecipeHandler");
+        }
+        if (clazz.isInterface()) {
+            throw new IllegalArgumentException("Class " + clazz.getName() + " is an interface and cannot be annotated with @IRecipeHandler.For");
```
I think in the ZenClassRegistry we had something similar wtih `getImplementationsOf` or something.
Check there to see that you haven't missed a check (maybe you can even reuse code from there lol).

Also check for abstract?

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/RecipeHandlerRegistry.java:
```diff
@@ -0,0 +1,89 @@
+            throw new IllegalArgumentException("Class " + clazz.getName() + " is an interface and cannot be annotated with @IRecipeHandler.For");
+        }
+        //noinspection RedundantCast
+        Arrays.stream(clazz.getAnnotationsByType(IRecipeHandler.For.class))
```
Same question as earlier, please check again with the "container" annotation to make sure all Annotations are found even if using the container directly?

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/RecipeHandlerRegistry.java:
```diff
@@ -0,0 +1,89 @@
+        //noinspection RedundantCast
+        Arrays.stream(clazz.getAnnotationsByType(IRecipeHandler.For.class))
+                .map(IRecipeHandler.For::value)
+                .filter(it -> (Class<?>) it != IRecipe.class)
```
Not necessary, but:
```suggestion
                .filter(IRecipe.class::equals)
```

TheSilkMiner replied:
This is the opposite of the check being performed

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/RecipeHandlerRegistry.java:
```diff
@@ -0,0 +1,89 @@
+            
+            if (target.getSuperclass() != null) classes.offer(target.getSuperclass());
+            Arrays.stream(target.getInterfaces()).forEach(classes::offer);
+        }
```
@jaredlll08 Remind me, did we want to check supertypes or not?

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/ReplacementRule.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.Objects;
+
+// TODO("Make this more than a data class")
+public final class ReplacementRule {
```
Throwing this up to discussion:
Should this be exposed to ZC?

TheSilkMiner replied:
Do you have an idea on how a replacement rule could be exposed? Because currently, the generic type used may not be even exposed to ZenCode, nor is `Optional`

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/commands/CTCommands.java:
```diff
@@ -114,6 +118,88 @@ public static void init(CommandDispatcher<CommandSource> dispatcher) {
+            player.world.getRecipeManager().recipes.forEach((iRecipeType, map) -> {
+                
+                AtomicBoolean handled = new AtomicBoolean(false);
+                IRecipeManager recipeManager = RecipeTypeBracketHandler.getRecipeManager(iRecipeType.toString());
```
Please check again that this doesn't throw an error?

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/commands/CTCommands.java:
```diff
@@ -114,6 +118,88 @@ public static void init(CommandDispatcher<CommandSource> dispatcher) {
+                StringBuilder builder = new StringBuilder("Recipe type: `").append(recipeManager.getCommandString())
+                        .append("`\n");
+                
+                map.values().stream().sorted((o1, o2) -> {
```
?
```java
sorted(Comparator.comparing(recipe -> regipe.getSerializer().getRegistryName())
        .thenComparing(IRecipe::getId))
```

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -0,0 +1,113 @@
+            if (newRecipeMaybe.isPresent()) {
+                return Optional.of(new ActionReplaceRecipe(manager, recipe, newRecipeMaybe.get(), rules));
+            }
+        } catch (final UnsupportedOperationException e) {
```
If you want to do it like that, please make a 
```java
public class ReplacementNotSupportedException extends UnsupportedOperationException{}
```

Or something with a better name

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -0,0 +1,113 @@
+    }
+    
+    @ZenCodeType.Method
+    public static Replacer forRecipes(final WrapperRecipe... recipes) {
```
What happens when a recipe is specified multiple times here?

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -0,0 +1,113 @@
+    }
+    
+    @ZenCodeType.Method
+    public static Replacer forTypes(final IRecipeManager... managers) {
```
What happens when a manager is specified multiple times here?

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -0,0 +1,113 @@
+        this.rules.add(new ReplacementRule(from, to));
+        return this;
+    }
+    
```
Maybe add a method 
```java
public Replacer replace(ReplacementRule rule) {
    this.rules.add(rule);
    return this;
}
```
Just so that other mods could ZenExpand the replacer class and add custom Replacer rules.
Like, for example, Mekanism could do something like
```less
Replacer.forTypes(craftingTable).replaceGas(<gas:oxygen> * 200, <gas:nitrogen> * 100).execute();
```
Or something (I have actually no idea if there are mek recipes that need a container with a gas present lol)
Or if you exposed the Rules to ZC, then even players could get creative lol.

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -0,0 +1,113 @@
+        final List<ReplacementRule> rules = Collections.unmodifiableList(this.rules);
+
+        Stream.concat(this.streamManagers(), this.streamRecipes())
+                .map(pair -> this.execute(pair.getFirst(), pair.getSecond(), rules))
```
In regards to the questions earlier (duplicates)
.distinct before map?

TheSilkMiner replied:
Mangers and recipes are now stored in a `Set`, so this should guarantee the lack of duplicates

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/handlers/crafttweaker/CTShapedRecipeHandler.java:
```diff
@@ -0,0 +1,36 @@
+                new MCItemStackMutable(recipe.getRecipeOutput()).getCommandString(),
+                IntStream.range(0, recipe.getRecipeHeight())
+                        .mapToObj(y -> IntStream.range(0, recipe.getRecipeWidth())
+                                .mapToObj(x -> ingredients.get(y * recipe.getRecipeWidth() + x))
```
Why not just add getters for `CTRecipeShaped#getIngredients()` which returns the stored `IIngredient[][]`?
Then you could also make this a bit easier, like
```java
Arrays.stream(recipe.getCrTIngredients())
    .map(row -> Arrays.stream(row).map(IIngredient::toSCommandString).collect(Collectors.joining(", ", "[", "]")
   .collect(Collectors.joining(", ", "[", "]")
```

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/handlers/crafttweaker/CTShapelessRecipeHandler.java:
```diff
@@ -0,0 +1,29 @@
+                StringUtils.quoteAndEscape(recipe.getId()),
+                new MCItemStackMutable(recipe.getRecipeOutput()).getCommandString(),
+                recipe.getIngredients().stream()
+                        .map(IIngredient::fromIngredient)
```
Why not add a getter for the IIngredient[] already stored in CTRecipeShapeless?

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/handlers/vanilla/CookingRecipeHandler.java:
```diff
@@ -0,0 +1,77 @@
+    }
+    
+    private static final Map<IRecipeType<?>, Pair<String, CookingRecipeFactory<?>>> LOOKUP = ImmutableMap
+            .<IRecipeType<?>, Pair<String, CookingRecipeFactory<?>>>builder()
```
You really need to be explicit here?

TheSilkMiner replied:
Yes, Java doesn't infer the types correctly otherwise

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/handlers/vanilla/CookingRecipeHandler.java:
```diff
@@ -0,0 +1,77 @@
+    
+    private static final Map<IRecipeType<?>, Pair<String, CookingRecipeFactory<?>>> LOOKUP = ImmutableMap
+            .<IRecipeType<?>, Pair<String, CookingRecipeFactory<?>>>builder()
+            .put(IRecipeType.BLASTING, Pair.of("blastFurnace", BlastingRecipe::new))
```
Do you really need that name here, can't you just use the `IRecipeManager#commandString` later?

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/handlers/vanilla/StoneCutterRecipeHandler.java:
```diff
@@ -0,0 +1,45 @@
+    public Optional<StonecuttingRecipe> replaceIngredients(final IRecipeManager manager, final StonecuttingRecipe recipe, final List<ReplacementRule> rules) {
+        
+        final Ingredient originalInput = recipe.getIngredients().get(0);
+        /*mutable*/ Ingredient input = originalInput;
+        for (final ReplacementRule rule : rules) {
+            if (rule.shouldReplace(input)) {
+                input = rule.getVanillaTo();
+            }
+        }
+    
+        if (originalInput == input) return Optional.empty();
```
It seems like this logic is used in quite a few of these handlers, maybe it makes sense to move it into an own method?

Maybe something like
```java
protected function replaceSingleIngredient(Ingredient originalInput, List<ReplacementRule> rules) {
    /*mutable*/ Ingredient input = originalInput;
    for (final ReplacementRule rule : rules) {
        if(rule.shouldReplace(input) {
            input = rule.getVanillaTo();
        }
    }
    return input;
}
```

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/ReplacementRule.java:
```diff
@@ -0,0 +1,67 @@
+        return this.from.asVanillaIngredient();
+    }
+    
+    public Ingredient getVanillaTo() {
+        return this.to.asVanillaIngredient();
+    }
+    
+    public boolean shouldReplace(final IIngredient ingredient) {
+        return this.from.contains(ingredient);
+    }
+    
+    public boolean shouldReplace(final Ingredient ingredient) {
+        return this.shouldReplace(IIngredient.fromIngredient(ingredient));
+    }
```
Maybe instead something like
```java
public Ingredient getReplacementFor(Ingredient ingredient){}
```
If you want it a bit more dynamic?

TheSilkMiner committed:
Make replacement rule generic and an interface

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add helper for ingredient replacement

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add getters for CT ingredients

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add simple first example of replacement rule

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Remove generification of recipe handlers

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add additional replacement rule implementations for common situations

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/IRecipeHandler.java:
```diff
@@ -0,0 +1,265 @@
+                        Optional.empty(),
+                        (optional, rule) -> rule.getReplacement(optional.orElse(ingredient), type),
+                        (oldOptional, newOptional) -> newOptional.isPresent()? newOptional : oldOptional
+                );
```
Why as stream and not as for-loop here?
May make it a bit easier than working with the reduce 😛

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -0,0 +1,140 @@
+    }
+    
+    @ZenCodeType.Method
+    public static Replacer forAllExcluding(final IRecipeManager... managers) {
```
`forAllTypesExcluding`?

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -0,0 +1,140 @@
+    }
+    
+    @ZenCodeType.Method
+    public Replacer replace(final IItemStack from, final IItemStack to) {
```
To could be IIngredient?

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -0,0 +1,140 @@
+    }
+    
+    @ZenCodeType.Method
+    public Replacer replace(final IIngredient from, final IItemStack to) {
```
To could IIngredient?

TheSilkMiner committed:
Fix generic over-engineering

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/handlers/vanilla/CookingRecipeHandler.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public String dumpToCommandString(final IRecipeManager manager, final AbstractCookingRecipe recipe) {
+        
+        return String.format(
+                "%s.addRecipe(%s, %s, %s, %s, %s);",
+                LOOKUP.get(recipe.getType()).getFirst(),
+                StringUtils.quoteAndEscape(recipe.getId()),
+                new MCItemStackMutable(recipe.getRecipeOutput()).getCommandString(),
+                IIngredient.fromIngredient(recipe.getIngredients().get(0)),
+                recipe.getExperience(),
+                recipe.getCookTime()
+        );
```
We had the discussion a while back:
Do we want a custom builder, since the conversions to the output format is in many cases
```java
return RecipeStringBuilder.forFormat("{manager}.addRecipe({name}, {output}, {input}, {experience}, {cookTime})")
    .set("manager", LOOKUP.get(recipe.getType()).getFirst())
    .quoteString("name", recipe.getId())
    .setItemStack("output", recipe.getRecipeOutput())
    .setIngredient("input", recipe.getIngredients().get(0)
    .set("experience", recipe.getExperience())
    .set("cookingTime", recipe.getCookTime())
    .toString();
```

TheSilkMiner committed:
Address some comments

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add replacement support to all remaining recipes

Also move common utility method to ReplacementHandlerHelper

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/ReplacementHandlerHelper.java:
```diff
@@ -0,0 +1,90 @@
+            this.delegate.defaultReturnValue(null);
+        }
+        
+        T get(final int index, final NonNullList<T> original) {
```
Why `NonNullList<T>` instead of just `List<T>` ?

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/ReplacementHandlerHelper.java:
```diff
@@ -0,0 +1,90 @@
+    
+    private ReplacementHandlerHelper() {}
+    
+    public static <T extends IRecipe<?>> Optional<T> replaceIngredients(final NonNullList<Ingredient> originalIngredients, final List<IReplacementRule> rules,
```
Why not `List<>` ?

kindlich replied:
Why only for Ingredient?
(Generic and Class<T> parameter?

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/handlers/crafttweaker/CTShapedRecipeHandler.java:
```diff
@@ -0,0 +1,68 @@
+    
+        final NonNullList<Ingredient> ingredients = recipe.getIngredients();
+        return String.format(
+                "craftingTable.addShaped(%s, %s, %s);",
```
Should we print some info if the recipe has a recipeFunction?

TheSilkMiner replied:
Added a print for this case. The format is still up for discussion.

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/handlers/crafttweaker/CTShapelessRecipeHandler.java:
```diff
@@ -0,0 +1,42 @@
+    public String dumpToCommandString(final IRecipeManager manager, final CTRecipeShapeless recipe) {
+    
+        return String.format(
+                "craftingTable.addShapeless(%s, %s, %s);",
```
Should we print something if the recipe has a function?

TheSilkMiner replied:
Added a print for this case. The format is still up for discussion.

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -0,0 +1,140 @@
+    }
+    
+    @ZenCodeType.Method
+    public void execute() {
```
Make sure this is only executed on the Server?

TheSilkMiner committed:
Add missed replacement for smithing recipes

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Rework ReplacementHandlerHelper to be more generic

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
kindlich, please tell me this is a good way of using actions

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Fix reduce not correctly combining results

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Move to an IAction for replacement

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/IRecipeHandler.java:
```diff
@@ -0,0 +1,230 @@
+     * class.
+     */
+    default Optional<T> replaceIngredients(final IRecipeManager manager, final T recipe, final List<IReplacementRule> rules) throws ReplacementNotSupportedException {
+        throw new ReplacementNotSupportedException("Replacement is not supported for this recipe class");
```
Maybe add the recipe manager name?
Also, you _could_ put the message string already in the exception lol

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/wrappers/WrapperRecipe.java:
```diff
@@ -67,4 +68,9 @@ public IItemStack getIcon() {
+    @ZenCodeType.Method
+    public void replace(final IIngredient from, final IIngredient to) {
+        Replacer.forRecipes(this).replace(from, to).execute();
+    }
```
Is there an easy way to also do 
```java
@Getter("commandString")
public String getCommandString() {
    //Get manager and print for this recipe
}
```

kindlich commented on src/main/java/com/blamejared/crafttweaker/api/recipes/ReplacementHandlerHelper.java:
```diff
@@ -0,0 +1,101 @@
+        
+        if (replacements.isEmpty()) return Optional.empty();
+        
+        final U[] newIngredients = arrayOf(ingredientClass, originalIngredients.length);
```
Why not `originalIngredients.clone()` ? 
Or `java.utils.Arrays.copyOf(originalIngredients, originalIngredients.length)`?

Then you would not need to hand over the class but could be sure that it's always a proper type.
It would also safe you the effort of re-adding the old ingredients lol.

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -0,0 +1,187 @@
+@ZenCodeType.Name("crafttweaker.api.recipes.Replacer")
+@Document("vanilla/api/recipes/Replacer")
+public final class Replacer {
+    private static final class ReplacerAction implements IRuntimeAction {
```
Move that to another file?

TheSilkMiner committed:
Stop warnings being printed for special recipes

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add rename functionality to Replacer

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Fix ingredients being discarded

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Move name autogeneration and fixing into NameUtils

Also rework the custom replacer to avoid double functions, since we were
going to compute it anyway to check for autogenerated names and avoid
fixing them if unnecessary, without allowing stuff to sneak past

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Allow usage of <p> and {@code} in Javadoc comments

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Refactor commands

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Allow expansions to add new rules

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add commandString for recipes

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add information regarding custom recipe functions

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Tweak exception message

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Fix some more documentation issues

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add documentation

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add user-defined exclusions

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Add automatic exclusions for unreplaceable recipes

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/replacement/ReplacerAction.java:
```diff
@@ -81,6 +85,20 @@ public boolean validate(final ILogger logger) {
+                " excluding {%s%s%s}",
+                userExcludedNames.isEmpty()? "" : userExcludedNames.stream().collect(Collectors.joining("\", \"", "\"", "\"")),
+                userExcludedNames.isEmpty() || defaultExcludedNames.isEmpty()? "" : ", ",
+                defaultExcludedNames.isEmpty()? "" : defaultExcludedNames.stream().collect(Collectors.joining("\" (automatic), \"", "\"", "\" (automatic)"))
```
I am not particularly fond of `(automatic)` as a marker. Does anyone have a better idea?

TheSilkMiner commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/replacement/ReplacerAction.java:
```diff
@@ -62,9 +68,7 @@ public String describe() {
-                this.exclusions.isEmpty()? "" : this.exclusions.stream()
-                        .map(ResourceLocation::toString)
-                        .collect(Collectors.joining("\", \"", " excluding {\"", "\"}"))
+                this.joinBothLists(this.userExclusions, this.defaultExclusions)
```
Adding this makes the line become even longer than it already is. Would it be better to split it into multiple lines? If so, which format would be better?

kindlich replied:
You do know that StringBuilders and the like exist?

TheSilkMiner committed:
Document GatherReplacementExclusionEvent

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

kindlich commented on Crafttweaker_Annotation_Processors/src/main/java/com/blamejared/crafttweaker_annotation_processors/processors/document/conversion/converter/comment/ParagraphMarkdownConverter.java:
```diff
@@ -0,0 +1,9 @@
+public class ParagraphMarkdownConverter {
+    
+    public String convertParagraphToMarkdown(final String docComment) {
+        return docComment.replace("</p>", "").replace("<p>", "").replace("<br />", "\n");
```
Well, can't you have these in normal markdown?

TheSilkMiner replied:
You can, but they break the markdown rendering IIRC, meaning that if you have `<p>Hello [this is a link](https://www.google.com)</p>`, then it breaks.
`<br />`s are not an issue, but I added them for consistency.

jaredlll08 commented on src/main/java/com/blamejared/crafttweaker/api/recipes/GatherReplacementExclusionEvent.java:
```diff
@@ -0,0 +1,76 @@
+ *
+ * <p>Mods should add recipes to this event <strong>if and only if</strong> replacement is actually impossible, and not
+ * based on whether replacement is supported on their end or not. As an example, {@code minecraft:armor_dye} is a recipe
+ * that is hardcoded and thus cannot have its ingredients replaced: this is a good candidate to addition.</p>
```
```suggestion
 * that is hardcoded and thus cannot have its ingredients replaced: this is a good candidate to exclude.</p>
```

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/commands/crafttweaker/HelpCommand.java:
```diff
@@ -0,0 +1,64 @@
+        
+        //Actually show the commands
+        for(int i = startCommandIndex; i < endCommandIndex; i++) {
+            final com.blamejared.crafttweaker.impl.commands.CommandImpl command = commands.get(keys.get(i));
```
Isn't this the same type as the already imported one?

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/commands/crafttweaker/InventoryCommands.java:
```diff
@@ -0,0 +1,64 @@
+            return 0;
+        }));
+    
+        CTCommands.registerCommand("inventory", CTCommands.playerCommand("tags", "Outputs the tags of the item in your inventory", (player, stack) -> {
```
```suggestion
        CTCommands.registerCommand("inventory", CTCommands.playerCommand("tags", "Outputs the tags of the items in your inventory", (player, stack) -> {
```

Plural?

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/commands/crafttweaker/InventoryCommands.java:
```diff
@@ -0,0 +1,64 @@
+                final String inventoryContents = IntStream.range(0, inventory.getSlots())
+                        .mapToObj(inventory::getStackInSlot)
+                        .filter(it -> !it.isEmpty())
+                        .map(it -> Pair.of(new MCItemStackMutable(it).getCommandString(), ItemTags.getCollection().getOwningTags(it.getItem())))
```
I always confuse them, is this the correct TagCollection (vs. the one from TagCollectionManager...)?
Also, you could directly use `TagManagerItem.INSTANCE.getAllTagsFor(it)`.
Then you'd at least make sure to be on the same TagCollection as `<tagmanager:items`

TheSilkMiner replied:
It is. Changed for consistency though.

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/commands/crafttweaker/RecipeCommands.java:
```diff
@@ -0,0 +1,89 @@
+        
+        CraftTweakerAPI.logInfo("Dumping all recipes that output %s!", new MCItemStackMutable(stack).getCommandString());
+        
+        final IItemStack workingStack = new MCItemStackMutable(stack).setAmount(1);
```
Remind me, the `stack` is that an actual reference to the player's held item, or a copy?
If the actual item, wouldn't this reduce their stacksize to 1?

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/brackets/RecipeTypeBracketHandler.java:
```diff
@@ -53,6 +57,17 @@ public static IRecipeManager getCustomManager(ResourceLocation location) {
+    public static IRecipeManager getOrDefault(final ResourceLocation location) {
+        return registeredTypes.computeIfAbsent(location, it -> {
+            final IRecipeType<?> type = Registry.RECIPE_TYPE.getOrDefault(location);
+            return type == null? null : new RecipeManagerWrapper(type);
```
Would this return a recipeManager for Script Recipes?

TheSilkMiner replied:
Yes in theory, although everything that calls this method explicitly checks against `crafttweaker:scripts`. Added one more check just in case, though

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/commands/crafttweaker/RecipeCommands.java:
```diff
@@ -0,0 +1,89 @@
+                .filter(filter)
+                .sorted(Comparator.comparing(RecipeCommands::serializer).thenComparing(IRecipe::getId))
+                .map(it -> dump(manager, it))
+                .peek(it -> builder.append(it).append('\n'))
```
Why not just a Join?
Isn't peek rather a "debugging method" lol

You could use (only a suggestion, not required)
```java
final String joined = ...
.map(it-> dump(manager, it))
.collect(Collectors.joining("\n");

if(joined.isEmpty()) {
    log("No recipe found");
} else {
    log("Recipe Type: " + manager.commandString + "\n" + join);
```

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/replacement/DefaultExclusionReplacements.java:
```diff
@@ -0,0 +1,26 @@
+                    .map(WrapperRecipe::getRecipe)
+                    .filter(it -> it instanceof SpecialRecipe)
+                    .forEach(event::addExclusion);
+        }
```
Just in case also exclude everything from the scripts manager, if it were to come up here?

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -1,132 +1,112 @@
-            return Optional.empty();
-        }
-    }
+    private static final Supplier<Function<ResourceLocation, ResourceLocation>> DEFAULT_REPLACER = Lazy.concurrentOf(
```
Why the 2nd order function?
Like, what advantage does having the Supplier bring instead of using the function directly over here?

TheSilkMiner replied:
Originally, this was so the default replacer would be initialized only if needed; with the changes made, though, there is effectively none. I'll remove it

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -157,31 +165,223 @@ public static Replacer forAllTypesExcluding(final IRecipeManager... managers) {
+                newName,
+                (fixed, mistakes) -> CraftTweakerAPI.logWarning(
+                        "%sInvalid recipe rename '%s' from '%s', mistakes:\n%s\nThe new rename '%s' will be used",
+                        position == null ? "" : position + ": ",
```
Check the implementation of `getZCScriptPositionFromStackTrace` again, but I think it never returned null but rather `CodePosition.UNKNOWN`

TheSilkMiner committed:
Address comments

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Hopefully make ReplacerAction#describe clearer

I've been told there's a thing called **variables**, but I am still not
sure what people are saying... I mean, the only true variables are
function parameters, aren't they? Because, that's the only thing that
can vary... everything else is constant

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

pupnewfster commented on src/main/java/com/blamejared/crafttweaker/api/recipes/GatherReplacementExclusionEvent.java:
```diff
@@ -0,0 +1,76 @@
+ * allowed to modify, meaning that users will have to remove and add recipes manually instead of being able to resort to
+ * automated tools.</p>
+ *
+ * <p>Mods should add recipes to this event <strong>if and only if</strong> replacement is actually impossible, and not
+ * based on whether replacement is supported on their end or not. As an example, {@code minecraft:armor_dye} is a recipe
+ * that is hardcoded and thus cannot have its ingredients replaced: this is a good candidate to exclude.</p>
```
I think it would be helpful to have this also mention the fact that CrT automatically adds special recipes to being excluded so that people don't think they need to include their regular crafting table special recipes.

pupnewfster commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/replacement/DefaultExclusionReplacements.java:
```diff
@@ -0,0 +1,34 @@
+            manager.getAllRecipes()
+                    .stream()
+                    .map(WrapperRecipe::getRecipe)
+                    .filter(it -> it instanceof SpecialRecipe)
```
Should this be checking `SpecialRecipe` like it is or do what vanilla does and check `IRecipe#isDynamic`. And should it check all the vanilla recipe types if so, as strictly speaking at least based on what I have seen in the recipe book vanilla allows for other recipe types to have "special" variants as well (even though there likely isn't much of a case there would be any, but if say some mod were to add a "special" furnace recipe then it would be able to be automatically added here.

TheSilkMiner committed:
Move from checking dynamicity of the recipe rather than classes

Thank you pupnewfster for the tip

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

kindlich commented on src/main/java/com/blamejared/crafttweaker/impl/recipes/Replacer.java:
```diff
@@ -0,0 +1,389 @@
+                        this.targetedManagers,
+                        this.targetedRecipes,
+                        Collections.unmodifiableList(this.rules),
+                        this.targetedManagers.stream().flatMap(manager -> DEFAULT_EXCLUSIONS.computeIfAbsent(manager, this::gatherDefaultExclusions).stream()).collect(Collectors.toList()),
```
Well, it is a bit of an edge-case, but, when two recipes with the same name in different managers exist:
1) Are both excluded even if only one manager was told to exclude the recipe
2) If both managers said to exclude it, this list would contain both names, so the user would see the recipe excluded two times

Just a heads-up, but can stay as-is

TheSilkMiner replied:
1. Two managers cannot reasonably define a recipe with the same name due to how datapacks are structured. Unless we are also considering recipes we add, but those would never be in the default exclusion list.
2. Right, I might convert that to a `Set` actually, since we are only expecting `Collection`s in the action anyway, and this automatically gets rid of duplicates.
